runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -59,16 +250,175 @@ extern "C" {
+            case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT64:
+                if (gprIndex <= maxGPRIndex) {
+                    thunkMem += STG(thunkMem, gprIndex, 0, R4, offsetFromR4 + (currArgIndex * 8));
+                    currArgIndex += 1;
+                }
+                gprIndex += 1;
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+                if (fprIndex <= maxFPRIndex) {
+                    thunkMem += STEY(thunkMem, fprIndex, 0, R4, offsetFromR4 + (currArgIndex * 8));
+                    currArgIndex += 1;
+                }
+                fprIndex += 2;
+		        if (i < 3) {
+		            gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+		        }
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+                if (fprIndex <= maxFPRIndex) {
+                    thunkMem += STDY(thunkMem, fprIndex, 0, R4, offsetFromR4 + (currArgIndex * 8));
+                    currArgIndex += 1;
+                }
+                fprIndex += 2;
+		        if (i < 3) {
+		            gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+		        }
+                break;
+            default:
+                Assert_VM_unreachable();
+        }
+
+    }
+
+    // Now jump to the common dispatcher helper routine.
+    thunkMem += IIHF(thunkMem, R1, ((reinterpret_cast<uint64_t>(metaData)) >> 32));
+    thunkMem += IILF(thunkMem, R1, reinterpret_cast<uint64_t>(metaData));
+    thunkMem += LAY(thunkMem, R2, 0, R4, offsetFromR4); // argument pointer
+    uint64_t upcallAddress = *(reinterpret_cast<uint64_t *>(reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher) + 8));
+    // Materialize the destination address in R6.
+    thunkMem += IIHF(thunkMem, R6, upcallAddress >> 32);
+    thunkMem += IILF(thunkMem, R6, upcallAddress);
+
+    // Now generate the branch to the helper
+    thunkMem += BASR(thunkMem, R7, R6);
+    thunkMem += BCR(thunkMem, 0, 0); // Mandatory NOP instruction
+    // Upon returning copy back the structure, then restore the stack pointer.
+    thunkMem += LMG(thunkMem, R6, R15, R4, 2064); // 1808+256
+    thunkMem += LAY(thunkMem, R4, 0, R4, 256);
+    // Then branch back
+    thunkMem += BC(thunkMem, 0xF, 0, R7, 2);
+
+    // Finish up before returning
+    vmFuncs->doneUpcallThunkGeneration(metaData, (void *)(metaData->thunkAddress));
+    metaData->functionPtr[0] = (UDATA)(*(reinterpret_cast<uint64_t *>(metaData->upCallCommonDispatcher)));
```

Comment:
it appeared you are not taking advantages of the fact that this env-value is passed back into the thunk call, since you are materializing metaData pointer into R1 anyway.